### PR TITLE
BigFloat compatibility for univariate optimization, fixes #271

### LINKED ISSUE #271: typecasts needed for BigFloat compatibility
The Optim.Brent() and Optim.GoldenSection() methods currently don't work with BigFloat.

The problem is very easy to fix: rel_tol and abs_tol need typecasts at the very end, inside the  UnivariateOptimizationResults() constructor.

### PR DESCRIPTION
**Pull request by andrewclausen**
The univariate optimization code unintentionally assumed that the choice
variable was a Float64.  Specifically, the interfaces accommodated an
AbstractFloat, but then would crash when reporting the results because of
missing typecasts.

This patch introduces approriate typecasts and test cases.

### CODE REVIEW

**Comment by pkofod** [comment]
Thanks for making this PR. However, I think the better thing to do is to drop the Float64 type in the results type. You can either change it to `::T` or just remove it altogether. If you pass in bigfloats, I don't think you would want a Float64 result back.

File: src/brent.jl [hunk 1]
```diff
@@ -162,8 +162,8 @@ function optimize{F <: Function, T <: AbstractFloat}(
                                          iteration,
                                          iteration == iterations,
                                          converged,
-                                         rel_tol,
-                                         abs_tol,
+                                         Float64(rel_tol),
```
Comment (pkofod) [inline src/brent.jl]: Revert these and remove the `::Float64` in the results type in `src/types.jl`

File: src/golden_section.jl [hunk 2]
```diff
@@ -103,8 +103,8 @@ function optimize{F<:Function, T <: AbstractFloat}(f::F, x_lower::T, x_upper::T,
                                          iteration,
                                          iteration == iterations,
                                          converged,
-                                         rel_tol,
-                                         abs_tol,
+                                         Float64(rel_tol),
```
Comment (pkofod) [inline src/golden_section.jl]: and these

File: src/optimize.jl [hunk 3]
```diff
@@ -261,7 +261,7 @@ function optimize{F<:Function, T <: AbstractFloat}(f::F,
     if show_trace
         @printf "Iter     Function value   Gradient norm \n"
     end
-    optimize(f, Float64(lower), Float64(upper), method;
+    optimize(f, T(lower), T(upper), method;
```
Comment (pkofod) [inline src/optimize.jl]: simply remove `T` here, as it's already there in the method signature

File: test/brent.jl [hunk 4]
```diff
@@ -5,4 +5,11 @@ let
         @assert Optim.converged(results)
         @assert norm(Optim.minimizer(results) - prob.minimizers) < 1e-7
     end
+
```
Comment (pkofod) [inline test/brent.jl]: I think it's simpler if we have 
```julia
for (name, prob) in Optim.UnivariateProblems.examples
    for T in (Float64, BigFloat)
        results = optimize(prob.f, convert(Array{T}, prob.bounds)..., method = Brent())

        @assert Optim.converged(results)
        @assert norm(Optim.minimizer(results) - prob.minimizers) < 1e-7
    end
end
```
than repeating the test loops twice

File: src/brent.jl (hunk 1 above)
Comment (andrewclausen) [inline src/brent.jl]: What about consistency between UnivariateOptimizationResults and MultivariateOptimizationResults?

Is there a good reason why MultivariateOptimizationResults uses ::Float64 for x_tol et al?

File: src/optimize.jl (hunk 3 above)
Comment (andrewclausen) [inline src/optimize.jl]: Oops, I meant to typecast rel_tol and abs_tol.

**Comment by andrewclausen** [comment]
I made all the changes you requested, but I don't know how to make a new Pull Request (or revise the old one).  Can you pull?

(I'm new to github, but I've been using git for 10 years.)

**Comment by KristofferC** [comment]
Pushing to the same branch (like you seem to have done here) will automatically update the Pull Request.

File: src/brent.jl (hunk 1 above)
Comment (pkofod) [inline src/brent.jl]: Nope, those should be removed as well. However, I am not entirely certain that all multivariate solvers will have methods for BigFloat from a-z, so maybe that can be left for another PR.

**Comment by pkofod** [comment]
Everything looks fine. Thanks! AppVeyor is failing for another reason that I have yet to fix (some barrier function issue on 32bit windows).

**Comment by pkofod** [comment]
Thanks for addressing this issue.
